specs/dwritecore/DWriteCore.md

```diff
@@ -0,0 +1,255 @@
+and enables rendering glyphs to the bitmap. On Windows, the object encapsulates a GDI memory DC with a GDI
+device-dependent bitmap (DIB) selected into it. However, on non-Windows platforms, there is no GDI DC,
+and the bitmap is just a system memory array. DWriteCore introduces a new version of the bitmap render target
+interface that an application can use to access the bitmap data without going through GDI.
+
+# Examples
+
+## Creating a restricted factory
+
+An application calls the [DWriteCreateFactory](https://docs.microsoft.com/en-us/windows/win32/api/dwrite/nf-dwrite-dwritecreatefactory)
+function to create a factory object, which is the entry-point to all other DirectWrite APIs. An application can
+create a _restricted factory_ if it does not want the factory or objects created from it to read from or write
+to any cross-process or persistent cache. In addition, a restricted factory's system font collection only includes
+well-known system fonts.
+
+The following example creates a restricted factory.
+
+```c++
+ComPtr<IDWriteFactory7> spFactory;
+HRESULT hr = DWriteCreateFactory(
+    DWRITE_FACTORY_TYPE_RESTRICTED,
+    __uuidof(IDWriteFactory7),
+    (IUnknown**)&spFactory
+    );
+```
+
+## Drawing glyphs to a system memory bitmap
+
+An application creates a bitmap render target by calling [IDWriteGdiInterop::CreateBitmapRenderTarget](https://docs.microsoft.com/en-us/windows/win32/api/dwrite/nf-dwrite-idwritegdiinterop-createbitmaprendertarget).
+A bitmap render target encapsulates a bitmap in system memory, and eanbles rendering glyphs to the bitmap. On Windows,
+the bitmap is actually a GDI DIB selected into a memory HDC, and it is possible to get the bitmap pixels using GDI
+functions. However, DWriteCore introduces a simpler method through the new IDWriteBitmapRenderTarget2 interface.
+The following example shows both the old way and the new way.
+
+```c++
+DWRITE_BITMAP_DATA_BGRA32 TextRenderer::GetBitmapData(IDWriteBitmapRenderTarget* renderTarget)
+{
+    DWRITE_BITMAP_DATA_BGRA32 result = {};
+
+    // Query the bitmap render target for the new interface.
+    ComPtr<IDWriteBitmapRenderTarget2> renderTarget2;
+    HRESULT hr = renderTarget->QueryInterface(&renderTarget2);
+    if (SUCCEEDED(hr))
+    {
+        // IDWriteBitmapRenderTarget2 exists, so we can get the bitmap the easy way.
+        hr = renderTarget2->GetBitmapData(OUT &result);
+        if (FAILED(hr))
+        {
+            throw Exception{ hr };
+        }
+    }
+    else
+    {
+        // We're using an older version that doesn't implement IDWriteBitmapRenderTarget2,
+        // so we have to get the bitmap by going through GDI. First get the bitmap handle.
+        HDC hdc = renderTarget_->GetMemoryDC();
+        HGDIOBJ dibHandle = GetCurrentObject(hdc, OBJ_BITMAP);
+        if (dibHandle == nullptr)
+        {
+            throw Exception{ HRESULT_FROM_WIN32(GetLastError()) };
```

Comment:
can we use wil directly from ProjectReunion? would simplify much of this COM and exception work vs. recreating similar types.

Reply:
Yes, we can use the [public WIL](https://github.com/microsoft/wil)

Reply:
Yes, it's open for anybody to use. Similar conversation with the regular inbox API reviews, though; "you need this other thing to use these APIs" is confusing unless we have some blanket statement.

Reply:
CONSIDER: Use `wil` to simplify windows-specific samples, but may not cross-compile to clang/gcc (@DefaultRyan, @dunhor, fyi).

DISCUSSION: cppwinrt does have support for exceptions with `check_result` but no equivalent to resource managers like `HGDIOBJ`.

Reply:
Leaving as-is because there's no resource management here (so raw HGDIOBJ works fine) and check_result is not applicable to a null check.